pkg/controllers/provisioningv2/rke2/machineprovision/controller.go

```diff
@@ -354,33 +357,39 @@ func (h *handler) OnRemove(key string, obj runtime.Object) (runtime.Object, erro
 		return obj, generic.ErrSkip
 	}
 
-	clusterName := infraObj.meta.GetLabels()[capi.ClusterLabelName]
+	clusterName := infra.meta.GetLabels()[capi.ClusterLabelName]
 	if clusterName == "" {
 		return obj, fmt.Errorf("error retrieving the clustername for machine, label key %s does not appear to exist for dynamic machine %s", capi.ClusterLabelName, key)
 	}
 
-	machine, err := rke2.GetMachineByOwner(h.machines, infraObj.meta)
-	if err != nil && !errors.Is(err, rke2.ErrNoMachineOwnerRef) {
+	machine, err := rke2.GetOwnerCAPIMachine(obj, h.machineCache)
+	if apierrors.IsNotFound(err) {
+		logrus.Debugf("[machineprovision] %s/%s: waiting: machine to be set as owner reference", infra.meta.GetNamespace(), infra.meta.GetName())
+		h.EnqueueAfter(infra, 10*time.Second)
+		return obj, generic.ErrSkip
+	}
+	if err != nil {
+		logrus.Errorf("[machineprovision] %s/%s: error getting machine by owner reference %v", infra.meta.GetNamespace(), infra.meta.GetName(), err)
 		return obj, err
 	}
 
-	if machine == nil || machine.Status.NodeRef == nil {
+	if machine.Status.NodeRef == nil {
 		// Machine noderef is nil, we should just allow deletion.
-		logrus.Debugf("[MachineProvision] There was no associated K8s node with this etcd dynamicmachine %s. Proceeding with deletion", key)
-		return h.doRemove(infraObj)
+		logrus.Debugf("[machineprovision] There was no associated K8s node with this machine %s. Proceeding with deletion", key)
+		return h.doRemove(infra)
 	}
 
-	cluster, err := h.rancherClusterCache.Get(infraObj.meta.GetNamespace(), clusterName)
+	cluster, err := h.rancherClusterCache.Get(infra.meta.GetNamespace(), clusterName)
 	if err != nil && !apierrors.IsNotFound(err) {
 		return obj, err
 	}
 	if apierrors.IsNotFound(err) || !cluster.DeletionTimestamp.IsZero() {
-		return h.doRemove(infraObj)
+		return h.doRemove(infra)
 	}
 
 	removed := true
 	// In the event we are removing an etcd node (as indicated by the etcd-role label on the node), we must safely remove the etcd node from the cluster before allowing machine deprovisioning
-	if val := infraObj.meta.GetLabels()["rke.cattle.io/etcd-role"]; val == "true" {
```

Comment:
nit: I don't know who put this here (was it me?) but this should probably using a strconv tolower to handle cases where `true` is not all lowercase